eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,220 @@
+                return true;
+            }
+        }
+        // If no @ServiceClient annotated with this class, return false
+        return false;
+    }
+
+    /**
+     *  Checks for public or protected constructor for the service client class.
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param ctorToken the CTOR_DEF AST node
+     */
+    private void checkConstructor(DetailAST ctorToken) {
+        final DetailAST modifiersToken = ctorToken.findFirstToken(TokenTypes.MODIFIERS);
+        // find constructor's modifier accessibility, no public or protected constructor
+        final AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "@ServiceClient class should not have any public or protected constructor.");
+        }
+    }
+
+    /**
+     * Checks for public static method named 'builder'. Should avoid to use method name, 'builder'.
+     *
+     * @param methodDefToken the METHOD_DEF AST node
+     */
+    private void checkMethodName(DetailAST methodDefToken) {
+        final DetailAST methodNameToken = methodDefToken.findFirstToken(TokenTypes.IDENT);
+        if (!BUILDER.equals(methodNameToken.getText())) {
+            return;
+        }
+
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        // find method's modifier accessibility, should not have a public static method called 'builder'
+        final AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) && modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+            log(modifiersToken, "@ServiceClient class should not have a public static method named ''builder''.");
+        }
+    }
+
+    /**
+     * Checks for the variable field of the subclass of ServiceClient.
+     * These fields should be final because these classes supposed to be immutable class.
+     *
+     * @param objBlockToken the OBJBLOCK AST node
+     */
+    private void checkClassField(DetailAST objBlockToken) {
+        for (DetailAST ast = objBlockToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (TokenTypes.VARIABLE_DEF != ast.getType()) {
+                continue;
+            }
+            final DetailAST modifiersToken = ast.findFirstToken(TokenTypes.MODIFIERS);
+            //VARIABLE_DEF token will always MODIFIERS token. If there is no modifier at the variable, no child under
+            // MODIFIERS token
+            if (!modifiersToken.branchContains(TokenTypes.FINAL)) {
+                log(modifiersToken, String.format("The variable field ''%s'' of @ServiceClient should be final. The class annotated with @ServiceClient supposed to be immutable.",
```

Comment:
Rather than `@ServiceClient`, are you able to state the class name? Also, small wording suggestion:

```suggestion
                log(modifiersToken, String.format("The variable field ''%s'' of @ServiceClient should be final. Classes annotated with @ServiceClient are supposed to be immutable.",
```